exp/operator/controllers/shared.go

```diff
@@ -0,0 +1,72 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/cluster-api/api/v1alpha4"
+	operatorv1 "sigs.k8s.io/cluster-api/exp/operator/api/v1alpha4"
+	"sigs.k8s.io/cluster-api/util/conditions"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func preflightChecks(ctx context.Context, c client.Client, provider client.Object) (*v1alpha4.Condition, ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	log.V(4).Info("Performing preflight checks.")
+
+	providerGVK := provider.GetObjectKind().GroupVersionKind()
+
+	providerList := &unstructured.UnstructuredList{}
+	providerList.SetGroupVersionKind(providerGVK)
+
+	if err := c.List(ctx, providerList); err != nil {
+		return nil, ctrl.Result{}, errors.Wrap(err, "failed to list providers")
+	}
+
+	for _, p := range providerList.Items {
+		// Skip if provider in the list is the same as provider it's compared with.
+		if p.GetNamespace() == provider.GetNamespace() && p.GetName() == provider.GetName() {
+			continue
+		}
+
+		preflightFalseCondition := conditions.FalseCondition(
+			operatorv1.PreflightCheckCondition,
+			operatorv1.MoreThanOneExistsReason,
+			v1alpha4.ConditionSeverityWarning,
+			operatorv1.MoreThanOneExistsMessage,
+		)
+
+		// CoreProvider is a singleton resource, more than one instances should not exist
+		if p.GetKind() == "CoreProvider" {
+			log.V(4).Info("More than one core provider exists in the cluster. Only one is allowed.")
+			return preflightFalseCondition, ctrl.Result{RequeueAfter: preflightFailedRequeueAfter}, nil
+		}
+
+		// For any other provider we should check that instances with similar name exist in any namespace
+		if p.GetKind() != "CoreProvider" && p.GetName() == provider.GetName() {
+			log.V(4).Info("More than one provider exists in the cluster. Only one is allowed.")
+			return preflightFalseCondition, ctrl.Result{RequeueAfter: preflightFailedRequeueAfter}, nil
+		}
+	}
+
+	return conditions.TrueCondition(operatorv1.PreflightCheckCondition), ctrl.Result{}, nil
```

Comment:
I don't think we should return changes to apply to the object as a variable, because this pattern quickly became hard to maintain as soon as the number of changes grow.
Let update the object inside this function instead

Reply:
I tried to update the function, but it is not trivial with unstructured.